src/compiler/moduleNameResolver.ts

```diff
@@ -1596,14 +1597,44 @@ function loadModuleFromFile(extensions: Extensions, candidate: string, onlyRecor
 function loadModuleFromFileNoImplicitExtensions(extensions: Extensions, candidate: string, onlyRecordFailures: boolean, state: ModuleResolutionState): PathAndExtension | undefined {
     // If that didn't work, try stripping a ".js" or ".jsx" extension and replacing it with a TypeScript one;
     // e.g. "./foo.js" can be matched by "./foo.ts" or "./foo.d.ts"
-    if (hasJSFileExtension(candidate) || (fileExtensionIs(candidate, Extension.Json) && state.compilerOptions.resolveJsonModule)) {
+    if (hasJSFileExtension(candidate)) {
         const extensionless = removeFileExtension(candidate);
         const extension = candidate.substring(extensionless.length);
         if (state.traceEnabled) {
             trace(state.host, Diagnostics.File_name_0_has_a_1_extension_stripping_it, candidate, extension);
         }
         return tryAddingExtensions(extensionless, extensions, extension, onlyRecordFailures, state);
     }
+    else if (
+        (extensions === Extensions.TypeScript || extensions === Extensions.DtsOnly) &&
+        hasExtension(candidate) &&
+        !fileExtensionIsOneOf(candidate, allSupportedExtensionsFlat) &&
+        !isDeclarationFileName(candidate)
+    ) {
+        const dotPosition = candidate.lastIndexOf(".");
+        const extensionless = candidate.substring(0, dotPosition);
+        const extension = candidate.substring(extensionless.length);
+        const newCandidate = `${extensionless}.d${extension}.ts`;
+        if (state.traceEnabled) {
+            trace(state.host, Diagnostics.File_name_0_has_a_1_extension_looking_up_2_instead, candidate, extension, newCandidate);
+        }
+        const result = tryFile(newCandidate, onlyRecordFailures, state);

```

Comment:
Is there a performance difference here with all of these added extension checks (I'm sure there are more than just here, but I might have missed it), or do they just not fire when there's not more than one dot?